### PR TITLE
[Dynamo x FSDP][4/x] Add support for params, buffers, submodules on FSDPManagedNNModuleVariable

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106890

fix2

cc @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/_dynamo/variables/nn_module.py [hunk 1]
```diff
@@ -801,3 +811,97 @@ def __init__(self, value, **kwargs):
         else:
             # this makes us behave like a usual UnspecializedNNModuleVariable for guarding purposes
             self.source = NotNNModuleSource(source)
+        self.module_key = module_key
+
+    def call_method(
```
Comment (wconstab) [inline torch/_dynamo/variables/nn_module.py:816]: i think overall, too much copy-paste to easily see the difference in behavior; refactoring could help reduce copy but also clarify what's unique here

File: torch/_dynamo/variables/builder.py [hunk 2]
```diff
@@ -793,8 +794,16 @@ def wrap_module(self, value: torch.nn.Module):
             #
             # ID_MATCH is required to disambiguate cases as simple as a unit test that constructs 2 models and wraps
             # them differently with different FSDP configs.  (test_dynamo_distributed.py -k test_fsdp_aot_eager)
+            base = self.name
```
Comment (wconstab) [inline torch/_dynamo/variables/builder.py:797]: what are the assumptions about self.name when we get to this line?  is it not assumed to be unique, therefore we try to make it so?  is this done in a way that's inconsistent with other dynamo constructions?

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -589,6 +594,25 @@ def register_attr_or_module(
...
+                    break
+                name = f"{base}_{i}"
+            options["guards"].add(source.make_guard(GuardBuilder.TYPE_MATCH))
+            options["guards"].add(source.make_guard(GuardBuilder.ID_MATCH))
+            vt = variables.nn_module.FSDPManagedNNModuleVariable(
+                target,
+                name,
+                **options,
+            )
+            return self.side_effects.track_object_existing(source, target, vt)
```
Comment (wconstab) [inline torch/_dynamo/output_graph.py:614]: is the pattern that top-level graph-input things come here to get registered/wrapped but other things go through builder?  I forgot the details, but its bugging me that we didn't have this case at all before and things seemed to mostly work.

**Review by wconstab — COMMENTED** [review]
some questions

File: torch/_dynamo/variables/builder.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/builder.py:797]: This is a pretty common pattern across the stack. The name here is just a result of source.name() but we want to give it a unique registration. Specialized modules already do this exact thing!

File: torch/_dynamo/variables/nn_module.py [hunk 4]
```diff
@@ -648,6 +653,11 @@ def __init__(self, value, **kwargs):
                 "ScriptModules aren't supported in UnspecializedNNModuleVariable"
                 " becuase their .forward function isn't a static member of their type"
             )
+        if (
+            getattr(value, "_is_fsdp_managed_module", False)
+            and type(self) == UnspecializedNNModuleVariable
+        ):
+            raise RuntimeError(f"Illegal construction {type(self)}")
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/variables/nn_module.py:660]: I think this is clear, but can elaborate.

File: torch/_dynamo/variables/nn_module.py [hunk 5]
```diff
@@ -801,3 +811,97 @@ def __init__(self, value, **kwargs):
...
+        key = self.module_key
+        options = VariableTracker.propagate(self, args, kwargs.values())
+
+        def assert_all_args_kwargs_const():
+            if not all(
+                x.is_python_constant() for x in itertools.chain(args, kwargs.values())
+            ):
+                raise unimplemented(f"non-const NNModule method {name}")
+
+        def get_kwargs(*names):
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/variables/nn_module.py:828]: Yea, I should dedupe

File: torch/_dynamo/variables/nn_module.py [hunk 6]
```diff
@@ -801,3 +811,97 @@ def __init__(self, value, **kwargs):
...
+            )
+
+        if name == "buffers":
+            return wrap_values(self.value.named_buffers(**get_kwargs("recurse")))
+        elif name == "named_buffers":
+            result = []
+            for name, buffer in self.value.named_buffers(
+                **get_kwargs("prefix", "recurse", "remove_duplicate")
+            ):
+                result.append(named_embed(name, buffer))
```
[the remark this answers is not in the document]
Comment (voznesenskym) [reply torch/_dynamo/variables/nn_module.py:889]: I think its the same, lemme dedupe

File: torch/_dynamo/output_graph.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:614]: Well, before, we would graph break on access here! It was never broken, just not covered.

**Comment by ezyang** [comment]
I don't have a good enough handle on FSDPManagedNNModuleVariable. @wconstab do you think you can do this entire review or should I also page this in?

**Comment by wconstab** [comment]
> I don't have a good enough handle on FSDPManagedNNModuleVariable. @wconstab do you think you can do this entire review or should I also page this in?

yea ill finish it today.

**Comment by wconstab** [comment]
@voznesenskym do you want me to re-review now or do you want me to wait till you get CI passing

File: torch/_dynamo/variables/nn_module.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_dynamo/variables/nn_module.py:765]: Refactored a lot

**Comment by voznesenskym** [comment]
Moved to https://github.com/pytorch/pytorch/pull/107923 due to new machine CLA issues

**Pull request closed without merging**